File: exercise/leetcode/0124/0124.py


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        max_val = [-1000]

        def dfs(r):
            if r is None:
                return 0

            l_p_val = max(0, dfs(r.left))
            r_p_val = max(0, dfs(r.right))
            cur_val = r.val
            max_val[0] = max(max_val[0], l_p_val+r_p_val+cur_val)

            return cur_val + max(r_p_val, l_p_val)

        dfs(root)
        return max_val[0]

File: exercise/leetcode/0124/0124.cc
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int max_val = -1000;
        dfs(root, max_val);
        return max_val;
    }

    int dfs(TreeNode* root, int& max_val) {
        if (!root) {
            return 0;
        }
        int lv = max(0, dfs(root->left, max_val));
        int rv = max(0, dfs(root->right, max_val));
        int cv = root->val;

        max_val = max(max_val, cv+lv+rv);
        
        return cv + max(lv, rv);

    }
};